Multi episode support

Bas-Man opened the pull request:
Refernces: #20 
- Update EPISODE Pattern removing need for `long` capture
- EPISODE will now also handle three digit episode numbers. (Rare but possible with shows that only have a single season)
- Add SECOND_EPISODE Pattern to catch EXXEXX
- Add last_episode() and is_multi_episode() functions. Naming convention to be decided.

Bas-Man committed:
remove the need for the `long` capture and capture two episodes.
Still need to solve the issue of extracting both captures...

Bas-Man committed:
remove the need for the `long` capture and capture two episodes.
Still need to solve the issue of extracting both captures...

Bas-Man committed:
Add support for accessing last_episode

Bas-Man committed:
Add support for accessing last_episode

Bas-Man committed:
Detect and store multi-episode information
 - update EPISODE Pattern removing need for `long` capture
 - EPISODE will now also handle three digit episode numbers. (Rare but 
possible with shows that only have a single season)
 - add SECOND_EPISODE Pattern to catch EXXEXX
 - add last_episode() and is_multi_episode() functions. Naming 
convention to be decided.

Bas-Man committed:
Detect and store multi-episode information
 - update EPISODE Pattern removing need for `long` capture
 - EPISODE will now also handle three digit episode numbers. (Rare but
possible with shows that only have a single season)
 - add SECOND_EPISODE Pattern to catch EXXEXX
 - add last_episode() and is_multi_episode() functions. Naming
convention to be decided.

Bas-Man commented:
Ideally I wanted to do the episode parse in one go. But I am not sure how some of the `check_pattern_and_extract` works. And it appears that I would not be able to pull multiple captures out in a single pass.

Variable names should be changed. Looking for guidance.
Would like some more test data to work with to make sure everything is ok and that nothing has been broken. I could revert back to the original `EPISODE` pattern now that its done in two passes.

IGI-111 commented on src/metadata.rs:
```diff
@@ -84,6 +85,9 @@ impl Metadata {
     pub fn episode(&self) -> Option<i32> {
         self.episode
     }
+    pub fn last_episode(&self) -> Option<i32> {
```
Probably merits some documentation of how and why that field exists for users.

Bas-Man replied:
Yes. Of course.

I am not happy with the actual naming. Any thoughts.

IGI-111 replied:
How about instead of this implied interval, a `fn multi_episode(&self) -> Option<Vec<i32>>`

Bas-Man replied:
I will take a look. But I was actually thinking that it might not be needed.
Also perhaps just `fn episodes(&self) -> Option<Vec<i32>>` making it plural should be fairly clear. Though I guess typing mistakes could cause issues...

I will come up with some sample code when I have a moment.

Bas-Man committed:
Document and change regex
- regex updated supports `E0Xe0X` and more `eXXEXXEXX....`
- Add tests
- Add documenation

Bas-Man committed:
Document and change regex
- regex updated supports `E0Xe0X` and more `eXXEXXEXX....`
- Add tests
- Add documenation

Bas-Man commented:
This might need an extra sanity check to ensure that last_episode is not less than episode. 

Eg:
`e03e0` or `e04e02` should set last_episode to `None` or return `Err()`

Have you ever seen `E04E02` where the episode numbering is in reverse order? Is this something that can safely be ignored as it is expected to never occur?

Bas-Man committed:
Add last_episode sanity check and additional tests
- Sanity check that last_episode does not get a Zero value if there is 
something like `E02e0`
- Additional checks for `E01E02` or `E01E03` where there is gap in the 
numbering sequence

Bas-Man committed:
Add last_episode sanity check and additional tests
- Sanity check that last_episode does not get a Zero value if there is 
something like `E02e0`
- Additional checks for `E01E02` or `E01E03` where there is gap in the 
numbering sequence

Bas-Man committed:
implement multi_episode()

Unable to avoid the return type being `Option<&Vec<i32>>`

Bas-Man committed:
implement multi_episode()

Unable to avoid the return type being `Option<&Vec<i32>>`

Bas-Man commented:
Could you take a look at my implementation for `multi_episode()`
I am not sure its the best. I was unable to avoid the reference. Thoughts?

Bas-Man committed:
Add more samples to simple.rs

Bas-Man committed:
Add more samples to simple.rs

IGI-111 commented on src/metadata.rs:
```diff
@@ -129,6 +153,34 @@ impl Metadata {
+    pub fn is_multi_episode(&self) -> bool {
+        self.last_episode.is_some()
+    }
+    pub fn multi_episode(&self) -> Option<&Vec<i32>> {
```
Pro tip: never return references to `Vec`, always return references to slices as `&Vec<i32>` implements `&[i32]` and will convert automatically.

Bas-Man replied:
I get the pro tip. So far I have not been able to avoid the `&Vec`
The compiler is complaining the vec is behind a borrow it's in and Option. If I am reading the compiler error correctly. Hence I am using the reference. 
I had to do the same thing in my own crate. Mmmm. Wonder if I am
Missing something.

Bas-Man replied:
I think perhaps. I just don't need the `Option<>`. Vecs can be empty.

IGI-111 replied:
You might be missing a `.transpose()` or something to bubble up the reference.

But you're right, we can do away with the `Òption` here and just use an empty vec on a `None`.

IGI-111 commented on src/metadata.rs:
```diff
@@ -12,6 +12,8 @@ pub struct Metadata {
     title: String,
     season: Option<i32>,
     episode: Option<i32>,
+    last_episode: Option<i32>,
+    multi_episode: Option<Vec<i32>>,
```
I don't think storing it twice is required, we can easily figure out the last episode from the Vec.

Bas-Man replied:
I was thinking last_episode could go away. Now that there is the `Vec`

Bas-Man replied:
last_episode removed
multi_episode renamed to `episodes` as it makes more sense and reads more naturally.

IGI-111 commented on src/metadata.rs:
```diff
@@ -158,13 +212,37 @@ impl FromStr for Metadata {
             },
         );
-
+        let episode_num = episode.map(|s| s.parse().unwrap());
+        // Only look for a last episode if one was found prevkously.
+        let mut last_episode_num: Option<i32> = None;
+        if let Some(first_episode) = episode {
+            last_episode = check_pattern_and_extract(
+                &pattern::LAST_EPISODE,
+                name,
+                &mut title_start,
+                &mut title_end,
+                |caps| caps.get(1).map(|m| m.as_str()),
+            );
+            // Sanity check that last_episode does not contain a value or 0 (Zero)
+            if let Some(l_episode) = last_episode {
+                if l_episode.len() == 1 && l_episode.contains('0') {
+                    last_episode = None;
+                } else {
+                    // Populate multi_episode
+                    let mut episode_numbers: Vec<i32> = Vec::new();
+                    for i in first_episode.parse().unwrap()..=l_episode.parse().unwrap() {
+                        episode_numbers.push(i);
+                    }
+                    multi_episode = Some(episode_numbers);
+                }
+            }
+            last_episode_num = last_episode.map(|s| s.parse().unwrap());
+        }
```
The logic looks correct to me, but the flow and naming could be much clearer. this should probably be a single functional expression that eventually resolves to an Option<i32>, not doing a bunch of conditional assigning.

Bas-Man replied:
I have made some improvements. At this stage I still need some conditionals. 

Will submit when I have some documentation done as well.

Bas-Man replied:
Submitted changes

Bas-Man committed:
refactor
- refactor code to create multi_episode
- remove last_episode()
- rename multi_episode -> episodes()
- update tests and simple.rs

Could remove is_multi_episode as this can be determined from episodes?

Bas-Man committed:
Merge in refactor code

Bas-Man committed:
remove option around Vec<i32>
- rename multi_episode to `episodes`
- move function just below `episode` for better ordering.
- update tests and documenation

Bas-Man committed:
Add previously removed mutli_epsode tests

Bas-Man commented on src/metadata.rs:
```diff
@@ -111,18 +111,25 @@ impl Metadata {
-    pub fn episodes(&self) -> Option<&Vec<i32>> {
-        self.multi_episode.as_ref()
+    pub fn episodes(&self) -> Vec<i32> {
+        self.episodes.clone()
```
I think this should return &Vec<i32> as this value will not change. Developers will just be borrowing the `Vec`

Bas-Man replied:
We can implement a  convenience function as so
```rust
pub fn last_episode(&self) -> Option<&i32> {
        match self.episodes.is_empty() {
            false => self.episodes.last(),
            true => None,
        }
}
```

In theory we could deprecate `episode`or implement in the same manner. Though I would suggest changing its name to `first_episode`

Just an idea.

IGI-111 replied:
I'd rather keep the interface least surprising, keep `episode` for most cases and provide `episodes` for the edge case so people can build whatever logic they want on top.

By the way, your snippet's `match` is useless. `self.episodes.last()` is equivalent to this since it's `None` on an empty slice. Or indeed `self.episodes().last()`.

Bas-Man replied:
Ah you are correct. last() returns an Option. So I can fix that.
Of course keep episode so as not to break existing code usage.

Bas-Man replied:
Agreed my code was redundant. I have addressed that in the final commit.
I have also ensured that there have been no user facing changes to the original code base. Everything works as previously.

Bas-Man committed:
update how episodes variable is populate
 - add convenience functions
 - add documenation
 - episode variable and usage remains the same.

Bas-Man committed:
Add CHANGELOG

Date format is based on Japanese standard. Please change to your
preferred format.

Bas-Man committed:
Add support to handle eXX-eXX or Exx eXX support space and dash

Bas-Man committed:
Add support to handle eXX-eXX or Exx eXX support space and dash

Bas-Man commented:
I am going to remove the first_ and last_ convenience methods. They can be implemented outside of the crate.

Sorry, I have caused some sort of offence.

Bas-Man committed:
remove is_multi_episode

Bas-Man committed:
Fix Changelog

IGI-111 commented on src/metadata.rs:
```diff
@@ -158,13 +189,38 @@ impl FromStr for Metadata {
+                &mut title_end,
+                |caps| caps.get(1).map(|m| m.as_str()),
+            );
+            match interim_last_episode {
```
That should be an `if let`

Bas-Man replied:
Switched to if let statement

IGI-111 commented on src/metadata.rs:
```diff
@@ -158,13 +189,38 @@ impl FromStr for Metadata {
+                    }
+                }
+                None => {}
+            } // End match interim_last_episode
```
Please don't mark block endings.

Bas-Man replied:
removed comments

IGI-111 commented on src/metadata.rs:
```diff
@@ -158,13 +189,38 @@ impl FromStr for Metadata {
+                }
+                None => {}
+            } // End match interim_last_episode
+        } // End Some(first_episode)
```
Same here.

Bas-Man replied:
Removed

IGI-111 reviewed (changes requested):
Sorry for the untimely reviews but I'm kinda busy with other projects.
I'll always come back to it eventually though.

Besides a few cosmetic changes this should be ready.

Please delete the changelog though. Not opposed to adding one down the line but I'm not going to maintain one before 1.0.0 since the interface isn't stable.

Bas-Man committed:
Requested changes
Remove CHANGELOG
Change match to if let statement

Bas-Man committed:
Merge branch 'master' into multi-episode-support

Bas-Man committed:
Requested changes

make extension case insensative
Add test for mixed case on extension

Bas-Man commented:
On a side note.
You can remove is_show() and is_special()
I have an idea of how to add this sort of support for my intended purpose. This way the core data gathering features only need to be implemented in this crate.
